File: frontend/apps/ios/Pantopus/Features/Homes/ResidencyReview/HomeResidencyReviewTransport.swift
import Foundation

@MainActor
protocol HomeResidencyReviewTransport {
    func read(scope: HomeResidencyReviewScope, claimId: String, sessionScope: String?) async throws -> HomeResidencyCurrentReview
    func decide(_ draft: PendingHomeResidencyReview, sessionScope: String) async throws -> HomeResidencyReviewReceipt
}

@MainActor
struct APIHomeResidencyReviewTransport: HomeResidencyReviewTransport {
    let api: APIClient

    func read(scope: HomeResidencyReviewScope, claimId: String, sessionScope: String?) async throws -> HomeResidencyCurrentReview {
        guard scope.isValid, HomePostalValidation.uuid(claimId) else { throw HomeResidencyReviewError.sessionChanged }
        let endpoint = Endpoint(
            method: .get,
            path: "/api/homes/\(scope.homeId)/claim/\(claimId)/review",
            headers: sessionScope.map { ["X-Pantopus-Session-Scope": $0] } ?? [:],
            cachePolicy: .reloadIgnoringLocalCacheData
        )
        let value: JSONValue
        do {
            value = try await api.request(endpoint)
        } catch APIError.unauthorized {
            throw HomeResidencyReviewError.sessionChanged
        } catch {
            throw HomeResidencyReviewError.unavailable
        }
        let review = HomeResidencyCurrentReview(value: value)
        guard review.matches(scope, claimId: claimId, expectedSession: sessionScope) else { throw HomeResidencyReviewError.unavailable }
        return review
    }

    func decide(_ draft: PendingHomeResidencyReview, sessionScope: String) async throws -> HomeResidencyReviewReceipt {
        guard draft.matches(draft.scope), HomeClaimReviewSnapshot.validToken(sessionScope) else { throw HomeResidencyReviewError.changed }
        let endpoint = Endpoint(
            method: .post,
            path: "/api/homes/\(draft.scope.homeId)/claim/\(draft.claimId)/\(draft.action.rawValue)",
            body: draft.body,
            headers: ["X-Pantopus-Session-Scope": sessionScope],
            cachePolicy: .reloadIgnoringLocalCacheData
        )
        let bytes: Data
        let status: Int
        do {
            let response = try await api.requestDataResponse(endpoint, includingForbidden: true, includingNotFound: true)
            bytes = response.data
            status = response.response.statusCode
        } catch let APIError.clientError(code, body) {
            guard let data = body?.data(using: .utf8) else { throw HomeResidencyReviewError.unknown }
            bytes = data
            status = code
        } catch APIError.unauthorized {
            throw HomeResidencyReviewError.sessionChanged
        } catch {
            throw HomeResidencyReviewError.unknown
        }
        guard let value = try? JSONDecoder().decode(JSONValue.self, from: bytes), let fields = value.dictValue else {
            throw HomeResidencyReviewError.unknown
        }
        if status != 200 { throw refusal(status: status, code: fields["code"]?.stringValue) }
        let review = HomeResidencyCurrentReview(value: value)
        guard review.matches(draft.scope, claimId: draft.claimId, expectedSession: sessionScope),
              fields["claim_id"]?.stringValue == draft.claimId,
              fields["target_id"]?.stringValue == review.applicantId,
              fields["action"]?.stringValue == draft.action.rawValue,
              fields["replayed"]?.boolValue != nil, let rawReceipt = fields["receipt"] else { throw HomeResidencyReviewError.unknown }
        let receipt = HomeResidencyReviewReceipt(value: rawReceipt)
        guard receipt.matches(draft) else { throw HomeResidencyReviewError.unknown }
        return receipt.projected()
    }

    private func refusal(status: Int, code: String?) -> HomeResidencyReviewError {
        let finalCodes = [
            "RESIDENCY_REVIEW_CHANGED",
            "CLAIM_NOT_PENDING",
            "OWNERSHIP_FLOW_REQUIRED",
            "MEMBERSHIP_RENEWAL_REQUIRED",
            "RESIDENCY_ROLE_FORBIDDEN",
            "PROPOSED_ROLE_FORBIDDEN",
            "PERMISSION_DELEGATION_FORBIDDEN"
        ]
        if [403, 409].contains(status), let code, finalCodes.contains(code) { return .refusal(code) }
        if [401, 403, 404].contains(status) || code == "SESSION_SCOPE_CHANGED" { return .unavailable }
        return .unknown
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/homes/HomeResidencyReviewTransport.kt
package app.pantopus.android.data.homes

import app.pantopus.android.data.api.services.HomeResidencyReviewApi
import kotlinx.coroutines.CancellationException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody

interface HomeResidencyReviewTransport {
    suspend fun read(
        scope: HomeResidencyReviewScope,
        claimId: String,
        session: String?,
    ): HomeResidencyCurrentReview

    suspend fun decide(
        draft: PendingHomeResidencyReview,
        session: String,
    ): String
}

class APIHomeResidencyReviewTransport(
    private val api: HomeResidencyReviewApi,
    private val codec: HomeResidencyReviewCodec,
) : HomeResidencyReviewTransport {
    override suspend fun read(
        scope: HomeResidencyReviewScope,
        claimId: String,
        session: String?,
    ): HomeResidencyCurrentReview {
        try {
            check(scope.isValid() && homeTaskUUID(claimId))
            val response = api.read(scope.homeId, claimId, session)
            val bytes = (response.body() ?: response.errorBody())?.use { it.string() }
            check(response.code() == HTTP_OK)
            return codec.current(checkNotNull(bytes), scope, claimId, session)
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            throw HomeResidencyReviewFailure(HomeResidencyReviewFailureKind.Unavailable)
        }
    }

    override suspend fun decide(
        draft: PendingHomeResidencyReview,
        session: String,
    ): String {
        try {
            check(codec.valid(draft, draft.scope) && HomeResidencyReviewCodec.reviewHash(session))
            val response =
                api.decide(
                    draft.scope.homeId,
                    draft.claimId,
                    draft.action.wire,
                    draft.requestJson.toRequestBody("application/json".toMediaType()),
                    session,
                )
            val json = (response.body() ?: response.errorBody())?.use { it.string() }
            val fields = codec.objectFrom(checkNotNull(json))
            if (response.code() != HTTP_OK) throw refusal(response.code(), fields["code"] as? String)
            val current = codec.current(fields, draft.scope, draft.claimId, session)
            check(fields["claim_id"] == draft.claimId && fields["target_id"] == current.applicantId)
            check(fields["action"] == draft.action.wire && fields["replayed"] is Boolean)
            return codec.receipt(codec.nested(fields["receipt"]), draft)
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (known: HomeResidencyReviewFailure) {
            throw known
        } catch (_: Exception) {
            throw HomeResidencyReviewFailure(HomeResidencyReviewFailureKind.Unknown)
        }
    }

    private fun refusal(
        status: Int,
        code: String?,
    ): HomeResidencyReviewFailure =
        HomeResidencyReviewFailure(
            when {
                status in FINAL_STATUSES && code in FINAL_CODES -> HomeResidencyReviewFailureKind.Refused
                status in DENIED_STATUSES || code == "SESSION_SCOPE_CHANGED" -> HomeResidencyReviewFailureKind.Unavailable
                else -> HomeResidencyReviewFailureKind.Unknown
            },
        )

    private companion object {
        const val HTTP_OK = 200
        val FINAL_STATUSES = setOf(403, 409)
        val DENIED_STATUSES = setOf(401, 403, 404)
        val FINAL_CODES =
            setOf(
                "RESIDENCY_REVIEW_CHANGED",
                "CLAIM_NOT_PENDING",
                "OWNERSHIP_FLOW_REQUIRED",
                "MEMBERSHIP_RENEWAL_REQUIRED",
                "RESIDENCY_ROLE_FORBIDDEN",
                "PROPOSED_ROLE_FORBIDDEN",
                "PERMISSION_DELEGATION_FORBIDDEN",
            )
    }
}
